R/art.con.internal.R

```diff
@@ -0,0 +1,320 @@
+    if (f.response != 0) {
+      stop("Formula must not have any variables on LHS (got ", variables[[1]], ").\n",
+           "Did you mean ", str_replace(f.orig.str, toString((variables[[1]])), ''))
+    }
+    
+    # char vector of names of rhs terms and their interactions
+    # e.g. ~ a:b:c -> c("a:b:c"))
+    term.labels = attr(f.terms, "term.labels")
+    
+    if (length(term.labels) != 1) {
+      stop("Model must have exactly one interaction and no other terms (" ,
+           length(term.labels), " terms given)")
+    }
+    
+    return(parse.art.con.string.formula(term.labels[[1]]))
+  } # end f is formula
+  
+  # looking for "a:b:c"
+  else if(is.character(f.orig) & length(f.orig) == 1){
+    return(parse.art.con.string.formula(f.orig))
+  }
+  
+  # Error if f is not formula or string
+  else{
+    stop("Contrast must either be formula of form ~ X1*X2*X3 or
+         term of form \"X1:X2:X3\")")
+  }
+}
+
+### Parses formula from supplied model
+### Validates that interaction term from f is in model formula
+### Does not validate model formula itself since model formula was validated when model was created
+### Given a formula like response ~ a*b*c + (1|d) + Error(g)
+### returns list with:
+### response: response (type name) (e.g. response)
+### fixed.variables: list of named fixed variables (of type name) (e.g. list(a, b, c)
+### grouping.variables: list of grouping variables (of type name) (e.g. list(1|d))
+### error.variables: list of error variables (of type name) (e.g. list(Error(g)))
+#' @importFrom plyr laply
+parse.art.model.formula = function(m.f, f.parsed){
+  
+  # get model formula terms
+  m.f.terms = terms(m.f)
+  # char vector of names of rhs terms and their interactions
+  # e.g. Response ~ a*b + (1|d) + Error(g) -> c("a", "b", "1 | d", "Error(g)", "a:b")) not necessarily in this order
+  m.f.term.labels = attr(m.f.terms, "term.labels")
+  
+  # check if interaction term from f is in model formula
+  f.interaction.term.label = f.parsed$interaction.term.label # f.parsed always has exactly one interaction
+  is.interaction.term = grepl(f.interaction.term.label, m.f.term.labels)
+  
+  # if interaction term from f is not in model formula, error
+  if(!any(is.interaction.term)){
+    stop("Term or formula passed to art.con or artlm.con must contain a single interaction term and no other terms, and the interaction term must be in art model formula.")
+  }
+  
+  # response ~ a*b*c*d + (1|d) + Error(g) -> list(response, a, b, c, d, 1|d, Error(g))
+  m.f.variables.all = as.list(attr(m.f.terms, "variables"))[c(-1)]
+  m.f.response = m.f.variables.all[[1]]
+  m.f.variables = m.f.variables.all[c(-1)]
+  
+  #determine which variables on the rhs are grouping variables, error variables, or fixed variables
+  is.grouping.variable = laply(m.f.variables, function(term) as.list(term)[[1]] == quote(`|`))
+  is.error.variable = laply(m.f.variables, function(term) as.list(term)[[1]] == quote(`Error`))
```

Comment:
should be able to use `sapply` instead of `plyr::laply` here (should be equivalent)